File: moorestech_client/Assets/Scripts/Client.Tests/WebUiHost/Train/WireContractTrainTimetableTest.cs
using System.Collections.Generic;
using System.IO;
using Client.WebUiHost.Common;
using Client.WebUiHost.Game.Topics;
using Client.WebUiHost.Game.Topics.BlockDetail;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using UnityEngine;

namespace Client.Tests.WebUiHost.Train
{
    // 時刻表取得状態のDTOとWireFixturesの一致を強制する
    // Enforce DTO ⇔ WireFixtures equality for the train inventory's timetable fetch state from the C# side
    public class WireContractTrainTimetableTest
    {
        [Test]
        public void LoadingTimetableFixtureMatchesDto()
        {
            var dto = CreateTrainInventory(TrainTimetableStateDto.Loading());
            dto.ItemSlots.Add(new BlockItemSlotDto { ItemId = 1, Count = 24 });
            dto.ItemSlots.Add(new BlockItemSlotDto { ItemId = 2, Count = 8 });
            AssertMatchesFixture(dto, "train_inventory.json");
        }

        // 取得済みのときだけ data が載る
        // data is carried only once the timetable is ready
        [Test]
        public void ReadyTimetableFixtureMatchesDto()
        {
            var position = new TrainStationPositionDto { X = 1, Y = 0, Z = 2 };
            var data = new TrainTimetableDto
            {
                TrainUnitId = "7",
                IsAutoRun = true,
                CurrentIndex = 0,
                Stops = new List<TrainTimetableStopDto> { new() { Position = position, Name = "North", Side = "front" } },
                Stations = new List<TrainTimetableStationDto> { new() { Position = position, Name = "North" } },
            };
            AssertMatchesFixture(CreateTrainInventory(TrainTimetableStateDto.Ready(data)), "train_inventory_timetable_ready.json");
        }

        [Test]
        public void UnavailableTimetableCarriesNoData()
        {
            var json = JObject.Parse(WebUiJson.Serialize(TrainTimetableStateDto.Unavailable()));
            Assert.That(json.ToString(Newtonsoft.Json.Formatting.None), Is.EqualTo("{\"kind\":\"unavailable\"}"));
        }

        private static BlockInventoryDto CreateTrainInventory(TrainTimetableStateDto timetable)
        {
            return new BlockInventoryDto
            {
                Open = true,
                Source = "train",
                BlockType = "Train",
                Identifier = "101",
                ItemSlots = new List<BlockItemSlotDto>(),
                FluidSlots = new List<BlockFluidSlotDto>(),
                Timetable = timetable,
            };
        }

        private static void AssertMatchesFixture(object dto, string fixtureName)
        {
            var actual = JToken.Parse(WebUiJson.Serialize(dto));
            var path = Path.Combine(Application.dataPath, "Scripts/Client.Tests/WebUi/WireFixtures", fixtureName);
            var expected = JToken.Parse(File.ReadAllText(path));
            Assert.IsTrue(JToken.DeepEquals(expected, actual), $"fixture mismatch: {fixtureName}\nexpected: {expected}\nactual: {actual}");
        }
    }
}

File: moorestech_web/webui/src/bridge/contract/wireContract.trainTimetable.test.ts
import { describe, it, expect } from "vitest";

import { parseTopicPayload } from "./validators";
import { loadFixture } from "./wireFixtures.test-helper";
import { Topics } from "../transport/protocol";

// 列車時刻表の取得状態の wire 契約。C# 側 WireContractTrainTimetableTest が同じフィクスチャを照合する
// Wire contract for the train timetable fetch state; the C# WireContractTrainTimetableTest matches the same fixtures
describe("train timetable fetch state (shared with C#)", () => {
  it("読み込み中と取得済みのフィクスチャを受理する", () => {
    expect(parseTopicPayload(Topics.blockInventory, loadFixture("train_inventory.json")).valid).toBe(true);
    expect(parseTopicPayload(Topics.blockInventory, loadFixture("train_inventory_timetable_ready.json")).valid).toBe(true);
  });

  it("取得状態ごとの形だけを受理する", () => {
    const train = loadFixture("train_inventory.json") as Record<string, unknown>;
    expect(parseTopicPayload(Topics.blockInventory, { ...train, timetable: { kind: "unavailable" } }).valid).toBe(true);
    // 取得済み以外は data を持てず、取得済みは data 必須、状態の省略は不可
    // Only ready may carry data, ready requires it, and the state cannot be omitted
    expect(parseTopicPayload(Topics.blockInventory, { ...train, timetable: { kind: "ready" } }).valid).toBe(false);
    expect(parseTopicPayload(Topics.blockInventory, { ...train, timetable: { kind: "loading", data: {} } }).valid).toBe(false);
    expect(parseTopicPayload(Topics.blockInventory, { ...train, timetable: undefined }).valid).toBe(false);
  });
});
